package.json

```diff
@@ -13,6 +13,14 @@
       "*": [
         "ts3.4/*"
       ]
+    },
+    "<4.7": {
+      "esm/*": [
+        "ts4.7/*"
```

Comment:
I think we can just have `ts3.4`, no? I mean change `"<4.0"` to `"<4.7"` at L9.

Reply:
I'm actually not really familiar with downlevel-dts but I'm guessing it should look like this...

```json
"typesVersions": {
  "<4.0": {
    "esm/*": [
      "ts3.9*"
    ],
    "*": [
      "ts3.9/*"
    ]
  },
  "<4.7": {
    "esm/*": [
      "ts4.6/*"
    ],
    "*": [
      "ts4.6/*"
    ]
  }
}
```

If we make it the following...

```json
"typesVersions": {
  "<4.7": {
    "esm/*": [
      "ts3.9*"
    ],
    "*": [
      "ts3.9/*"
    ]
  }
}
```

Then users with ts version 4.6 will have to use types of ts version 3.9 when they could have used types with ts version 4.6

Am I getting this right?

Also why do we have <4.0 in the first place? which 4.0 feature we're using? and why it was 3.4 instead of 3.9? Am I missing something?

If we're not using any 4.0 feature per se then I think it'd look like the latter version.

Reply:
> Then users with ts version 4.6 will have to use types of ts version 3.9 when they could have used types with ts version 4.6

Yes, would it be too much trouble?

> Also why do we have <4.0 in the first place? which 4.0 feature we're using?

I don't remember which, but I added it when I thought we used some 4.0 feature. The reason targeting 3.9 is it's the default minimum version that downlevel-dts supports.

So, we expect people use the latest typescript version, if it's not possible, we offer the lowest TS version as we can.

Reply:
> Yes, would it be too much trouble?

Yeah that would unnecessarily downlevel some features afaiu. For eg let's say we are using some feature from 4.5, the user is using 4.6, but if we downlevel <4.7 to 3.9 then it'll also downlevel the feature were using from 4.5 even if the user has support for it. So I'd say we should downlevel <4.7 to 4.6.

And by that same logic I'd suggest you to consider downleveling <4.0 to 3.9 instead of 3.4. And also consider not downleveling <4.0 at all, at the end of the day if users want they can downlevel themselves.

Reply:
> Yeah that would unnecessarily downlevel some features

I understand it, but not sure the real problem. It's just non-ideal.
For maintainability, I would like to have a straightforward config for all zustand/jotai/valtio.

> downleveling <4.0 to 3.9 instead of 3.4.

Would that break if people use 3.4? It's rare lately. But, I remember we had issues with old CRAv4 (especially with codesandbox back then), or something.

> at the end of the day if users want they can downlevel themselves.

Not sure if I follow this. How can users do it by themselves?

At this point, I still want it to be <4.7 to 3.4 for long-term simplicity and compatibility.

Reply:
> I understand it, but not sure the real problem. It's just non-ideal.

I'm not exactly sure how downlevel-dts works, for instance at some places we're using `[...A, ...B]` which is not available for 3.4 afaik, so I'm not sure what will downlevel-dts do with it (probably leave it as it is?). The point I'm trying to make is downlevel-dts will mess with our types uncessarily even if users have version as latest as 4.6, that could be a real problem.

> Not sure if I follow this. How can users do it by themselves?

By writing a postinstall script that overwrites .d.ts with downlevel-dts or something like that.

> Would that break if people use 3.4? It's rare lately. But, I remember we had issues with old CRAv4 (especially with codesandbox back then), or something.

It could I guess. That's another problem with `typesVersion`, like not only we have to care about our source code for breaking changes but also the `typesVersion`. Like say if we publish a version with some `typesVersion`, then removing it in the next versions could also be a breaking change.

Frankly speaking I personally wouldn't bother with downleveling my types if it were my package, it's user's responsibility to use the required version or downlevel themselves.

> At this point, I still want it to be <4.7 to 3.4 for long-term simplicity and compatibility.

I'm fine with that, just I think downleveling for users that use very recent ts versions if not latest is a bit problematic.

But anyway now that we have #1029 we don't need `out` annotation xD

Reply:
> By writing a postinstall script that overwrites .d.ts with downlevel-dts or something like that.

I see. As far as I recall, the reason we introduced downlevel-dts was codesandbox (or cra4) didn't support newer TS back then. So, `postinstall` wouldn't be a reasonable option.

> I think downleveling for users that use very recent ts versions if not latest is a bit problematic.

It makes sense, especially "very recent" part, and I get the point.

> But anyway now that we have #1029 we don't need `out` annotation and hence no downleveling xD

I'll check it. So, if we merge #1029, we don't need #1004?

Reply:
> I'll check it. So, if we merge #1029, we don't need #1004?

Yes in fact #1004 is now actually not the right fix, 4.8 changed between two version `-dev` versions as I mention in #1029. Also if ts fixes the possible bug I mention in #1029 then we won't even need #1029 xD